### PR TITLE
Cluster disaster recovery

### PR DESCRIPTION
**Pull request by gsanchietti**
# Changes

- Add `/usr/local/agent/bin/cluster-backup` binary used by `download-cluster-backup` cluster API
- Add `retrieve-cluster-backup` cluster API
- Add `restore-cluster` cluster API
- Add `restore-modules` cluster API
- Add `read-backup-repositories` cluster API
- Add `read-backup-snapshots` cluster API

Disaster recovery flow: 
1. `retrieve-cluster-backup`: download or upload the encrypted and compressed JSON dump
2. `restore-cluster`: create the cluster with data from the dump
3. `read-backup-repositories`: search for modules inside the backup
4. `restore-modules`: restore the selected modules

## download-cluster-backup

Dump backup configuration in JSON format. The dump is compressed using `gzip` and finally encrypted using `gpg`.

Execute a backup:
```
api-cli run download-cluster-backup
```

Response example:
```json
{"path": "ad40cfe5342d731cc4423daefc535e9489666a39264cf8659bc8a4a0e6c4f32e/dump.json.gz.gpg"}
```

The backup is encrypted with the passphrase contained inside `/var/lib/nethserver/cluster/state/backup/passphrase`.
The passphrase is a SHA256 sum.
If the above file is empty, the `cluster-backup` will use the SHA256 of the admin password from redis.

Conventions:
- the dump contains a `version` field that an be used to handle multiple type of backups during the restore
- the JSON `cluster` field mimic redis namespaces, `vpn` field contains Wireguard configuration, `modules` contains a map between module instances, labels and backups
- redis fields with set type are converted to lists

Generated `backup.gz` should be quite small, on a test machine, it is ~1.1KB.

Example:
```json
{
  "version": "1",
  "timestamp": 1644228067,
  "modules": {
    "cc7335d7-4d67-408c-8c35-42257667e51b": {
      "backups": [
        "1"
      ]
    }
  },
  "vpn": {
    "endpoint": "cs1.leader.cluster0.gs.nethserver.net:55820",
    "listen_port": "55820",
    "destinations": ""
  },
  "cluster": {
    "repository": {
      "default": {
        "url": "https://raw.githubusercontent.com/NethServer/ns8-repomd/repomd/",
        "status": "1",
        "testing": "1"
      }
    },
    "backup": {
      "1": {
        "enabled": "1",
        "repository": "e181c936-1bc3-5032-a809-d8b0551eebe9",
        "schedule_hint": "{\"custom\": \"\", \"interval\": \"daily\", \"minute\": \"0\", \"monthDay\": \"1\", \"time\": \"00:00\", \"weekDay\": \"sunday\"}",
        "retention": "5",
        "name": "Backup to B2 repo",
        "schedule": "*-*-* 00:00:00"
      }
    },
    "backup_repository": {
      "e181c936-1bc3-5032-a809-d8b0551eebe9": {
        "url": "b2:giacomons8",
        "password": "d49784aa7a9a22885354aaaeb8cbe581df5ec4f53280xxxxxxxxxxxxxxxxxxxxx",
        "name": "B2 repo",
        "provider": "backblaze",
        "b2_account_id": "003648087fdefxxxxxxxxxxx",
        "b2_account_key": "K003Frjo2poKxxxxxxxxxxxxxxxxxx"
      }
    },
    "user_domain": {
      "ldap": {
        "ns.test": {
          "conf": {
            "schema": "rfc2307",
            "bind_dn": "cn=test",
            "bind_password": "mypass",
            "base_dn": "dc=test",
            "tls": "on",
            "tls_verify": "on"
          },
          "providers": [
            "ldap.ns.test:389"
          ],
          "ui_names": {
            "ldap.ns.test:389": "myprovider"
          }
        }
      }
    },
    "network": "10.5.4.0/24",
    "ui_name": "mycluster1",
    "favorites": [
      "dokuwiki1"
    ]
  }
}
```

Not saved:
- UI users, roles and passwords
- Audit logs from SQLite database
- Traefik certificates

## retrieve-cluster-backup

The backup can be loaded inside the cluster from a URL or UI upload.

Retrieve backup from URL.
If `password` is empty, try to decode with sha254 of `admin` user from Redis, otherwise use the given password:
```
api-cli run retrieve-cluster-backup --data '{"type": "url", "url": "https://myserver/dump.json.gz.gpg", "tls_verify": true, "password": ""}'
```

## restore-cluster

After the backup has been retrieved, just execute:
```
api-cli run restore-cluster
```

## read-backup-repositories

List all backups inside all all backup repositories:
```
api-cli run read-all-backup-repositories
```

## read-backup-snapshots

List all snapshots inside a given backup. Example:
```
api-cli run read-backup-snapshots --data '{"path": "dokuwiki/dokuwiki1@cc7335d7-4d67-408c-8c35-42257667e51b", "repository": "e181c936-1bc3-5032-a809-d8b0551eebe9"}' 
```

Response example:
```
[
  {
    "timestamp": 1643964838,
    "id": "a2a4cb238e4bd428900756376d4ff94009a8f487effe25c145bfaffa72406693"
  },
  {
    "timestamp": 1644037213,
    "id": "e993a67283eefa7a5d148284790e526f70f360788ed71e17e18406bc5a8a1185"
  },
  {
    "timestamp": 1644123618,
    "id": "4ea007b7770cfab125ab6035ae5371fe0e87464b8caa1607b2ced8c7e8732b4a"
  },
  {
    "timestamp": 1644210013,
    "id": "72a3cecc8b6acbd1610ebabd9aae1ac2b0b96864fa986a7437188036378e61ad"
  },
  {
    "timestamp": 1644296416,
    "id": "ebd99130107f282e47e7eb161de8fe14e0b832b7eacb04c9bf60761407e6081a"
  }
]
```

## restore-modules

Restore a list of modules, example:
```
api-cli run bulk-restore-modules --data '[{"path": "dokuwiki/dokuwiki1@cc7335d7-4d67-408c-8c35-42257667e51b", "repository": "e181c936-1bc3-5032-a809-d8b0551eebe9"}]' 
```

### CODE REVIEW

**Comment by gsanchietti** [comment]
Encryption:
```
echo "mypass" > /etc/nethserver/backup.pass
/usr/local/agent/bin/cluster-backup | gzip | gpg --batch -c --pinentry-mode loopback --passphrase-file   /etc/nethserver/backup.pass > backup.gz.gpg
```

File: core/imageroot/usr/local/agent/bin/cluster-backup [hunk 1]
```diff
@@ -0,0 +1,130 @@
+#!/usr/local/agent/pyenv/bin/python3
```
Comment (gsanchietti) [inline core/imageroot/usr/local/agent/bin/cluster-backup:1]:
```suggestion
#!/usr/bin/env python3
```

File: core/imageroot/usr/local/agent/bin/cluster-backup [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+# if the passhprase does not exists, use the sha256 of admin password
+if os.path.isfile(f'{output_dir}/passphrase'):
+    subprocess.run(['gpg', '--batch', '-c', '--pinentry-mode', 'loopback', '--passphrase-file', f'{output_dir}/passphrase', f'{output_dir}/dump.json.gz'], check = True)
+else:
+    rdb = agent.redis_connect(host='127.0.0.1', privileged = True)
+    admin = rdb.acl_getuser('admin')
+    rdb.close()
+
+    fp = tempfile.NamedTemporaryFile(mode="w", delete=False)
```
Comment (gsanchietti) [inline core/imageroot/usr/local/agent/bin/cluster-backup:124]: Try with with, open as `w+`

File: core/imageroot/var/lib/nethserver/cluster/actions/bulk-restore-modules/50restore_modules [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import tempfile
+
+request = json.load(sys.stdin)
+
+# source
+for module in request:
+    data = module
+    data.setdefault('snapshot', '')
+    data.setdefault('node', 1)
+    agent.tasks.run_nowait(agent_id = 'cluster', action = 'restore-module', data = data,  endpoint = "redis://cluster-leader")
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/bulk-restore-modules/50restore_modules:37]:
```suggestion
    proc = agent.tasks.run(agent_id = 'cluster', action = 'restore-module', data = data,  endpoint = "redis://cluster-leader")
    if proc['exit_code'] > 0:
        errors = errors + 1
```

File: core/imageroot/var/lib/nethserver/cluster/actions/read-all-backup-repositories/50read [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        pchild = subprocess.Popen(rclone_cmd + ["REMOTE_PATH/" + oroot['Path']], **popen_args)
+        for backup in json.load(pchild.stdout):
+            # parse ISO 8601 date but remove ending Z (UTC timezone)
+            mod_time = datetime.fromisoformat(backup['ModTime'][0:-1])
+            mod_time = mod_time.replace(tzinfo=timezone.utc).timestamp()
+
+            if not '@' in backup["Path"] or backup["IsDir"] is False:
+                continue # skip non-repo entries
+            parts = backup['Path'].split('@')
+            backups.append({'name': oroot['Path'], 'path': f'{oroot["Path"]}/{backup["Path"]}', 'instance': parts[0], 'uuid': parts[1], 'timestamp': int(mod_time), 'repository_id' : repo_uuid, 'repository_name': repo['name'], 'repository_provider': repo['provider'], 'repository_url': repo['url']})
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/read-all-backup-repositories/50read:55]: Add the same info to read-backup-repository?

File: core/imageroot/var/lib/nethserver/cluster/actions/restore-cluster/30load [hunk 5]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/restore-cluster/30load:4]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

File: core/imageroot/var/lib/nethserver/cluster/actions/restore-cluster/20create [hunk 6]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/restore-cluster/20create:4]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

File: core/imageroot/var/lib/nethserver/cluster/actions/read-backup-snapshots/50read [hunk 7]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/read-backup-snapshots/50read:4]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

File: core/imageroot/var/lib/nethserver/cluster/actions/bulk-restore-modules/50restore_modules [hunk 8]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/bulk-restore-modules/50restore_modules:4]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

File: core/imageroot/var/lib/nethserver/cluster/actions/read-all-backup-repositories/50read [hunk 9]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/read-all-backup-repositories/50read:4]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

File: core/imageroot/var/lib/nethserver/cluster/actions/retrieve-cluster-backup/10validate [hunk 10]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/retrieve-cluster-backup/10validate:4]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

File: core/imageroot/var/lib/nethserver/cluster/actions/retrieve-cluster-backup/20save [hunk 11]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/retrieve-cluster-backup/20save:4]:
```suggestion
# Copyright (C) 2022 Nethesis S.r.l.
```

File: core/imageroot/var/lib/nethserver/cluster/actions/read-all-backup-repositories/50read (hunk 4 above)
Comment (gsanchietti) [reply core/imageroot/var/lib/nethserver/cluster/actions/read-all-backup-repositories/50read:55]: This API will probably totally replace the old `read-backup-repository`.
So we are going to remove `read-backup-repository` as soon as the UI has been adjusted to call `read-all-backup-repositories`

File: core/imageroot/usr/local/agent/bin/cluster-backup (hunk 2 above)
Comment (gsanchietti) [reply core/imageroot/usr/local/agent/bin/cluster-backup:124]: I couldn't make it work correctly, gpg complains with:
```
gpg: symmetric encryption of '/var/lib/nethserver/cluster/state/backup/dump.json.gz' failed: Invalid passphrase
```

I'm going to keep the current implementation.

/cc @DavidePrincipi

File: core/imageroot/var/lib/nethserver/cluster/actions/download-cluster-backup/60link [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    echo "Backup '$BACKUP' not found"
+    exit 1
+fi
+
+rm -rf $OUT_DIR
+sha=$(sha256sum -z $BACKUP | awk '{print $1}')
+mkdir -p $OUT_DIR/$sha/
+cp $BACKUP $OUT_DIR/$sha/
+
+echo "{\"path\": \"/cluster-admin/backup/$sha/dump.json.gz.gpg\"}"
```
Comment (DavidePrincipi) [inline core/imageroot/var/lib/nethserver/cluster/actions/download-cluster-backup/60link:36]: Leading URL path components should be added by the UI code, or rendered as a relative URL

```suggestion
echo "{\"path\": \"$sha/dump.json.gz.gpg\"}"
```

Comment (gsanchietti) [reply core/imageroot/var/lib/nethserver/cluster/actions/download-cluster-backup/60link:36]: With current implementation is easier to share the link.
Still I can change it easily.

@andre8244 are you ok if I change as requested?

File: core/imageroot/var/lib/nethserver/cluster/actions/retrieve-cluster-backup/20save [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        pg = subprocess.run(cmd + [fp.name, f'{output_dir}/dump.json.gz.gpg'], check = True)
+    finally:
+        os.remove(fp.name)
+
+# load json from gzip file
+fgz = gzip.open(f'{output_dir}/dump.json.gz', mode='r')
+dump = json.loads(fgz.read())
+fgz.close()
+
+result = {'timestamp': dump['timestamp'], 'cluster': dump['cluster']['ui_name'], 'modules': len(dump['modules']), "vpn": dump['vpn']['endpoint'], 'domains': len(dump['cluster']['user_domain']), 'backup_repositories': len(dump['cluster']['backup_repository'])}
```
Comment (gsanchietti) [inline core/imageroot/var/lib/nethserver/cluster/actions/retrieve-cluster-backup/20save:76]:
```suggestion
result = {'timestamp': dump['timestamp'], 'cluster': dump['cluster']['ui_name'] or '', 'modules': len(dump['modules']), "vpn": dump['vpn']['endpoint'], 'domains': len(dump['cluster']['user_domain']), 'backup_repositories': len(dump['cluster']['backup_repository'])}
```

File: core/imageroot/var/lib/nethserver/cluster/actions/download-cluster-backup/60link (hunk 12 above)
Comment (andre8244) [reply core/imageroot/var/lib/nethserver/cluster/actions/download-cluster-backup/60link:36]: No problem!

Comment (gsanchietti) [reply core/imageroot/var/lib/nethserver/cluster/actions/download-cluster-backup/60link]: Changed.

**Review by DavidePrincipi — APPROVED** [review]

**Pull request closed without merging**